kratos/python_scripts/python_solver.py

```diff
@@ -0,0 +1,175 @@
+        else:
+            raise Exception("Other model part input options are not yet implemented.")
+        KratosMultiphysics.Logger.PrintInfo("ModelPart", self.main_model_part)
+        KratosMultiphysics.Logger.PrintInfo("::[PythonSolver]:: ", "Finished reading model part.")
+
+    def PrepareModelPart(self):
+        """This function prepares the ModelPart for being used by the PythonSolver
+        """
+        pass
+
+    def GetMinimumBufferSize(self):
+        """This function returns the minimum buffer size needed for this PythonSolver
+        """
+        raise Exception('Please implement "GetMinimumBufferSize" in your derived solver')
+
+    def ImportModelPart(self):
+        warning_msg  = 'Using "ImportModelPart" is deprecated and will be removed in the future!\n'
+        warning_msg  = 'Use "ReadModelPart" + "PrepareModelPart" instead'
+        KratosMultiphysics.Logger.PrintWarning("::[PythonSolver]::", warning_msg)
+        self.ReadModelPart()
+        self.PrepareModelPart()
+
+    def ExportModelPart(self):
+        """This function exports the ModelPart to and mdpa-file
+        """
+        name_out_file = self.settings["model_import_settings"]["input_filename"].GetString()+".out"
+        file = open(name_out_file + ".mdpa","w")
+        file.close()
+        KratosMultiphysics.ModelPartIO(name_out_file, KratosMultiphysics.IO.WRITE).WriteModelPart(self.main_model_part)
+
+    def AdvanceInTime(self, current_time):
+        """This function advances the PythonSolver in time
+        Usage: It is designed to be called once per solution step, before performing the solution
+        """
+        pass
+
+    def Initialize(self):
+        """This function initializes the PythonSolver
+        Usage: It is designed to be called ONCE, BEFORE the execution of the solution-loop
+        """
+        pass
+
+    def Finalize(self):
+        """This function finalizes the PythonSolver
+        Usage: It is designed to be called ONCE, AFTER the execution of the solution-loop
+        """
+        pass
+
+    def Predict(self):
+        """This function performs all the required operations that should be executed
+        (for each step) ONCE, AFTER initializing the solution step.
+        """
+        pass
+
+    def InitializeSolutionStep(self):
+        """This function performs all the required operations that should be executed
+        (for each step) BEFORE solving the solution step.
+        """
+        pass
+
+    def FinalizeSolutionStep(self):
+        """This function performs all the required operations that should be executed
+        (for each step) AFTER solving the solution step.
+        """
+        pass
+
+    def SolveSolutionStep(self):
+        """This function solves the current step.
+        It can be called multiple times within one solution step
+        """
+        pass
+
+    def Check(self):
```

Comment:
I would add some trivial checks (like settings is a Parameter, input file exists, etc.)

Reply:
the type of the input is checked in the constructor 

I thought of this function to be called before the SolutionLoop (I will add it to the AnalysisStage)
=> It should call methods like Strategy.Check() to validate if the input is correct

Therefore it cannot validate if the input file has been found
(btw for this one the error is quite clear already:
`RuntimeError: Error: Error opening mdpa file : Example.mdpa`)